### ISSUE TITLE
Problem when using contravariant typevar and a chain of generic calls

### ISSUE THREAD

**Issue by matejcik** [microsoft/pyright#4006]
**Describe the bug**
Call to `restream(A(B()))` does not typecheck, but `x = A(B()); restream(x)` is fine.

**To Reproduce**
```python
import typing as t

A = t.TypeVar("A", contravariant=True)


class Poc(t.Generic[A]):
    value: A


def restream(subcon: Poc[A]) -> Poc[A]:
    return subcon


def optional(subcon: Poc[A]) -> Poc[A | None]:
    return Poc()


class Concrete(Poc[int]):
    pass


select: Poc[int | None] = Poc()
optional_val = optional(Concrete())

restream(optional(Concrete()))  # !!!! error here !!!!
restream(optional_val)  # no error
restream(select)  # no error

reveal_type(Optional(Concrete()))  # Poc[int | None]
reveal_type(optional_val)  # Poc[int | None]
reveal_type(select)  # Poc[int | None]
```
the error in question:
```
  /home/matejcik/projekty/prace/construct-classes/poc.py:31:19 - error: Argument of type "Concrete" cannot be assigned to parameter "subcon" of type "Poc[A@Optional]" in function "Optional"
    "Concrete" is incompatible with "Poc[int | None]"
      TypeVar "A@Poc" is contravariant
        Type "int | None" cannot be assigned to type "int"
          Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
```

It looks like pyright is incorrectly unifying the incoming and outgoing types of `optional()`?

**Expected behavior**
Both `restream(Optional(Concrete()))` and `restream(optional_val)` should have the same result. Hopefully they should type-check, because the code as written seems strictly correct.

**VS Code extension or command-line**
pyright 1.1.273, also in vscode extension

**Comment by erictraut** [comment]
Thanks for the bug report. Yes, this appears to be an issue with bidirectional type inference.

Here's a simplified code sample that exhibits the problem:

```python
from typing import Generic, TypeVar

T = TypeVar("T", contravariant=True)

class A(Generic[T]): ...

def func1(x: A[T]) -> A[T | None]: ...

x1: A[int | None] = func1(A[int]())
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.274, which I just published. It will also be included in a future release of pylance.
